modules/renter/contractor/contractor.go

```diff
@@ -81,41 +76,76 @@ func (c *Contractor) SetAllowance(a modules.Allowance) error {
 		return errors.New("renew window must be less than period")
 	}
 
-	err := c.formContracts(a)
+	// check that allowance is sufficient to store at least one sector
+	numSectors, err := maxSectors(a, c.hdb)
 	if err != nil {
 		return err
+	} else if numSectors == 0 {
+		return errInsufficientAllowance
 	}
 
-	// Set the allowance.
+	// prevent ProcessConsensusChange from renewing contracts until we have
+	// finished
+	c.contractLock.Lock()
+	defer c.contractLock.Unlock()
+
 	c.mu.Lock()
-	c.allowance = a
-	err = c.saveSync()
+	endHeight := c.blockHeight + a.Period
+
+	// check if allowance has different period or funds; if so, we should
+	// renew existing contracts with the new allowance.
+	// TODO: compare numSectors instead of allowance.Funds?
+	var renewSet []modules.RenterContract
+	if a.Period != c.allowance.Period || a.Funds.Cmp(c.allowance.Funds) != 0 {
+		for _, contract := range c.contracts {
+			renewSet = append(renewSet, contract)
+		}
+		// TODO: dangerous -- figure out a safer approach later
+		c.contracts = map[types.FileContractID]modules.RenterContract{}
```

Comment:
You can pretty easily atomically update the allowance. And you can also (more work, but easy work) easily keep track of all the in-progress rewnewals happening at the moment of an updated allowance. Then, when doing a renew, you can verify against the allowance that was atomically updated.

So the logic flow probably looks something like this:

1. atomically update allowance
2. all renewals go into a queue - both from changing the allowance and from ProcessConsensusChange, and upon being insert they check for conflicts (is there already a renewal for this contract in the queue, and if so it should be replaced)
3. some other thread pops stuff out of the queue and updates the set of contracts.

That should be safe